### PR TITLE
Remove nations iter

### PR DESCRIPTION
**Pull request by pranavtifr**
Remove the nations iterator. 
Test:

There should be no behavioral change.

### CODE REVIEW

File: client/dialogs.cpp [hunk 1]
```diff
@@ -566,15 +566,15 @@ void races_dialog::refresh()
...
+        if (!is_nation_playable(&pnation) || !is_nation_pickable(&pnation)
+            || !nation_is_in_group(&pnation, group)) {
+          continue;
+        }
+        count++;
       }
-      count++;
-    }
-    nations_iterate_end;
+    };
```
Comment (lmoureaux) [inline client/dialogs.cpp:577]: `std::count_if`

File: common/nation.cpp [hunk 2]
```diff
@@ -455,166 +455,80 @@ struct nation_type *nation_by_number(const Nation_type_id nation)
   if (nation < 0 || nation >= game.control.nation_count) {
     return nullptr;
   }
-  return nations + nation;
+  return &nations[0] + nation;
```
Comment (lmoureaux) [inline common/nation.cpp:458]: `&nations[nation]`

File: common/nation.cpp [hunk 3]
```diff
@@ -455,166 +455,80 @@ struct nation_type *nation_by_number(const Nation_type_id nation)
...
 {
-  int i;
-
-  nations = new nation_type[num];
   game.control.nation_count = num;
-
-  for (i = 0; i < num; i++) {
-    nation_init(nations + i);
-  }
+  nations.resize(game.control.nation_count);
```
Comment (lmoureaux) [inline common/nation.cpp:523]: `game.control.nation_count` => `nations.size()`? (possibly wrapped in a function)

**Review by lmoureaux — COMMENTED** [review]
Some suggestions

File: common/nation.cpp (hunk 2 above)
Comment (pranavtifr) [reply common/nation.cpp:458]: :smile:  Looking at C too long makes me forget the simplest of things xD

File: common/nation.cpp (hunk 3 above)
Comment (pranavtifr) [reply common/nation.cpp:523]: No, the vector is size 0 until it is resized here. I'd keep game.control.nation_count this way, atleast for now.

File: client/dialogs.cpp [hunk 4]
```diff
@@ -647,28 +640,28 @@ void races_dialog::set_index(int index)
...
-    }
-    item = new QTableWidgetItem;
-    selected_nation_tabs->insertRow(i);
-    auto s = get_nation_flag_sprite(tileset, pnation);
-    if (pnation->player) {
-      f = item->font();
-      f.setStrikeOut(true);
-      item->setFont(f);
+  for (const auto &pnation : nations) {
+    {
```
Comment (lmoureaux) [inline client/dialogs.cpp:644]: Double braces

File: client/tilespec.cpp [hunk 5]
```diff
@@ -1082,8 +1082,11 @@ bool tilespec_reread(const char *new_tileset_name,
   governments_iterate_end;
   extra_type_iterate(pextra) { tileset_setup_extra(tileset, pextra); }
   extra_type_iterate_end;
-  nations_iterate(pnation) { tileset_setup_nation_flag(tileset, pnation); }
-  nations_iterate_end;
+  for (auto &pnation : nations) {
+    {
```
Comment (lmoureaux) [inline client/tilespec.cpp:1086]: Double braces

File: common/actions.cpp [hunk 6]
```diff
@@ -588,19 +588,19 @@ static void hard_code_oblig_hard_reqs_ruleset()
...
-    if (nation_barbarian_type(pnation) == ANIMAL_BARBARIAN) {
-      oblig_hard_req_register(req_from_values(VUT_NATION, REQ_RANGE_PLAYER,
-                                              false, true, true,
-                                              nation_number(pnation)),
-                              true,
-                              N_("All action enablers for %s must require"
-                                 " a non animal player actor."),
-                              ACTRES_CONQUER_CITY, ACTRES_NONE);
+  for (const auto &pnation : nations) {
+    {
```
Comment (lmoureaux) [inline common/actions.cpp:592]: More double braces

**Review by lmoureaux — CHANGES_REQUESTED** [review]
Not having the extra indent from the double braces would make this PR much smaller

**Comment by pranavtifr** [comment]
Ah crap. Really stupid mistake. I should've checked before inlining the macros. Would have been easier then

**Comment by jwrober** [comment]
HOLDING POST BETA 3

**Review by jwrober — APPROVED** [review]
Ran an ASAN autogame. Looks good.

**Comment by jwrober** [comment]
Need @pranavtifr to mark the requested changes fixed or fix them before merge.

**Pull request merged**